### PR TITLE
Fix bug for cumsum on cupy chunked dask arrays

### PR DESCRIPTION
**Pull request by GenevieveBuckley**
Fixes https://github.com/dask/dask/issues/9315

And very likely, it fixes cumsum for a broader range of array types than just copy (but I haven't tested that here).

- [x] Closes https://github.com/dask/dask/issues/9315
- [x] Tests added / passed
- [x] Passes `pre-commit run --all-files`

### CODE REVIEW

**Comment by GenevieveBuckley** [comment]
FYI, I'm using [this trick Matt Rocklin suggests](https://stackoverflow.com/questions/50861900/including-keyword-arguments-kwargs-in-custom-dask-graphs) to insert a keyword argument into a dask task:

```python
from dask.utils import apply
task = (apply, func, args, kwargs)  # func(*args, **kwargs)
```

It's a useful trick, and one that should possibly be documented here too: https://docs.dask.org/en/stable/custom-graphs.html

**Review by jrbourbeau — COMMENTED** [review]
Thanks @GenevieveBuckley! Hope you don't mind, I merged `main` to resolves some unrelated CI failures. 

cc @pentschev who is both a `cupy` and NEP-35 person

File: dask/array/reductions.py [hunk 1]
```diff
@@ -1512,7 +1519,12 @@ def cumreduction(
     dsk = dict()
     for ind in indices:
         shape = tuple(x.chunks[i][ii] if i != axis else 1 for i, ii in enumerate(ind))
-        dsk[(name, "extra") + ind] = (np.full, shape, ident, m.dtype)
+        dsk[(name, "extra") + ind] = (
+            apply,
+            np.full,
+            (shape, ident, m.dtype),
+            {"like": x._meta},
```
[the remark this answers is not in the document]
Comment (pentschev) [reply dask/array/reductions.py:1526]: The `like=` kwarg in `np.full` (along with its counterparts `np.empty`, `np.ones`, `np.zeros`) has only been added for API completeness. Another valid approach here is to use `np.full_like` instead, this works on all versions that Dask supports today, all compatibility code has even been removed in https://github.com/dask/dask/pull/7939.

I think the equivalent code would look like the following:

```python
        dsk[(name, "extra") + ind] = (
            apply,
            np.full_like,
            (x._meta, ident, m.dtype),
            {"shape": shape},
        )
```

@GenevieveBuckley could you check if the above works?

File: dask/array/tests/test_reductions.py [hunk 2]
```diff
@@ -583,6 +583,15 @@ def test_array_cumreduction_out(func):
     assert_eq(x, func(np.ones((10, 10)), axis=0))
 
 
+@pytest.mark.parametrize("func", [np.cumsum, np.cumprod])
+def test_cumreduction_with_cupy(func):
```
[the remark this answers is not in the document]
Comment (GenevieveBuckley) [reply dask/array/tests/test_reductions.py]: Good point, done!

File: dask/array/reductions.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (GenevieveBuckley) [reply dask/array/reductions.py]: Ooh, good point. I hadn't considered that.

Yes, the alternative code @pentschev suggests using `full_like` works as expected. I've made the change.

**Comment by GenevieveBuckley** [comment]
Thanks Peter and James for the helpful review (especially for catching the numpy version compatibility concern). 

I've made the adjustments suggested. The new test passes on my local machine, so hopefully it will do the same on the GPU CI.

**Review by pentschev — APPROVED** [review]
Looks good to me, thanks @GenevieveBuckley !

**Comment by jsignell** [comment]
Seems like this is good to go!

**Pull request merged**